components/suggest/src/yelp.rs

```diff
@@ -142,299 +153,260 @@ impl SuggestDao<'_> {
+            let keyword = &query_string[0..keyword_lowercase.len()];
+            let count = keyword.split_whitespace().count();
+            return Ok(Some((
+                keyword.to_string(),
+                true,
+                query_words.get(count..).unwrap_or_default(),
+            )));
+        };
 
-        let mut post_modifier = None;
-        for n in (1..=MAX_MODIFIER_WORDS_NUMBER).rev() {
-            let mut candidate_chunks = words.rchunks(n);
-            let candidate = candidate_chunks.next().unwrap_or(&[""]).join(" ");
-            if self.is_modifier(&candidate, post_modifier_type)? {
-                post_modifier = Some(candidate);
-                break;
-            }
+        if query_string.len() < SUBJECT_PREFIX_MATCH_THRESHOLD {
+            return Ok(None);
         }
 
-        let mut without_modifiers = query;
-        if let Some(ref modifier) = pre_modifier {
-            without_modifiers = &without_modifiers[modifier.len()..];
-        }
-        if let Some(ref modifier) = post_modifier {
-            without_modifiers = &without_modifiers[..without_modifiers.len() - modifier.len()];
-        }
+        // Oppositely, this checks if the query is a substring of keyword.
+        if let Some(keyword_lowercase) = self.conn.try_query_one::<String, _>(
+            "SELECT keyword
+             FROM yelp_subjects
+             WHERE keyword BETWEEN :query AND :query || x'FFFF'
+             ORDER BY LENGTH(keyword) ASC, keyword ASC
+             LIMIT 1",
+            named_params! {
+                ":query": query_string.to_lowercase(),
+            },
+            true,
+        )? {
+            // Preserve the query as the user typed it including its case.
+            let keyword = format!(
+                "{}{}",
+                query_string,
+                &keyword_lowercase[query_string.len()..]
+            );
+
+            let count = std::cmp::min(query_words.len(), keyword.split_whitespace().count());
+            return Ok(Some((
+                keyword,
+                false,
+                query_words.get(count..).unwrap_or_default(),
+            )));
+        };
 
-        Ok((
-            without_modifiers.trim().to_string(),
-            pre_modifier,
-            post_modifier,
-        ))
+        Ok(None)
     }
 
-    /// Find the subject from the given string.
-    /// It returns the Option. If it is not none, it contains the tuple as follows:
+    /// Find the location sign for given query.
+    /// It returns Option<tuple> as follows:
     /// (
-    ///   String: Subject.
-    ///   bool: Whether the subject matched exactly with the parameter.
+    ///   String: The keyword in DB (but the case is inherited by query).
+    ///   &[&str]: Words after removed matching location sign.
     /// )
-    fn find_subject(&self, candidate: &str) -> Result<Option<(String, bool)>> {
-        if candidate.is_empty() {
+    fn find_location_sign<'a>(
```

Comment:
Oh also I forgot about `find_location_sign` -- could you add similar tests in `yelp.rs` for it too please?